### PR TITLE
LPD-28923 - LPP-48494 Dropdown fragment content overflows- edit summary Dropdown fragment content overflows

### PR DESCRIPTION
**Pull request by georgel-pop-lr**
Fixes both:

[LPD-28923 - LPP-48494 Dropdown fragment content overflows- edit summary Dropdown fragment content overflows](https://liferay.atlassian.net/browse/LPD-28923)
[LPD-28918 - LPP-48984 Dropdown fragment content overflows, part 2- edit summary Dropdown fragment content overflows](https://liferay.atlassian.net/browse/LPD-28918)

### CODE REVIEW

**Comment by liferay-echo** [comment]
Thank you for submitting this PR. 👏

Make sure it follows the [Echo Pull Request Template](https://gist.github.com/liferay-echo/5a4455285ff1800f9da146816d85e67d#file-pull-request-template-md) if you haven't already done so. 

Learn more about it [here](https://liferay.atlassian.net/wiki/spaces/ENGECHO/pages/2027455014/Echo+Pull+Request+Etiquette).

File: modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js [hunk 1]
```diff
@@ -170,6 +170,11 @@ function main() {
 	}
 	else {
 		toggle.addEventListener('click', handleToggleClick);
+		toggle.addEventListener('keydown', (event) => {
```
Comment (veroglez) [inline modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js:173]: Hi @sandrodw3, @georgel-pop-lr! I was just passing by here, I leave a couple of comments that I have seen

Here the `click` event should be fired when we press `enter`. I just saw that in the view mode it works well, but in edit mode it does not. Could keyboard navigation within the layout viewport be interfering here?

Also, in the dropdown button, can we also add an `aria-controls` associated with the `id` of the element it shows? just like the Clay component has

Screenshot from 2024-07-03 20-54-59

Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js:173]: Thanks, @veroglez, I think you meant is working when editing the page but not when you are viewing it.

I add a preventDefault, and it seems to work just fine, the strange thing is that I can't debug the code when on view mode, it shows me an empty page (I added the debugger; in handleToggleClick):
image

Is this happening for you as well?

Regarding aria-controls I will check it out.

Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js:173]: I added aria-controls as well, marking this as solved.

Comment (veroglez) [reply modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js:173]: @georgel-pop-lr, It seems that the problem comes here https://github.com/liferay/liferay-portal/blob/f87fa5dd559e0c59e48bd53e9f6f0209e7c6e043/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/hooks/app_hooks/useLayoutKeyboardNavigation.js#L81. From here, we manage the keyboard navigation in the layout. The `event.preventDefault()` is causing the click action of the button is not being executed. Actually this prevent default is there for the editables (when you press enter and edit it), so we can move the `preventDefault()` inside this conditional https://github.com/liferay/liferay-portal/blob/f87fa5dd559e0c59e48bd53e9f6f0209e7c6e043/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/hooks/app_hooks/useLayoutKeyboardNavigation.js#L99.
This way, the button will return to having its click event.

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+const ENTER_KEY = 'Enter';
+
+test('Check dropdown menu is displayed correctly in all resolutions', async ({
+	apiHelpers,
+	pageEditorPage,
+	site,
+}) => {
+	const dropdownId = getRandomString();
+
+	await test.step('Create dropdown fragment and change the configuration', async () => {
```
Comment (veroglez) [inline modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:43]: Analyzing the two issues, I think we could simplify these tests into a single test. Both issues try to position the dropdown, depending on the position it has within the layout. So the test could be:
- Create a page with a dropdown. You can directly create the fragment with the `d-flex` and `justify-content-end` classes adding this within the [fragment definition ](https://github.com/liferay/liferay-portal/blob/6d11491fa2624ee9a23dc3d27d9747aec75d60a3/modules/test/playwright/tests/layout-content-page-editor-web/utils/getFragmentDefinition.ts#L22) (we can add a new parameter to the method: `cssClasses`)
```
"cssClasses: ['d-flex', 'justify-content-end']" 
```
- Open the dropdown
- Verify that the dropdown has the correct position by checking that it has the inline style right
- Change the viewport to mobile, verify that it still has the same inline style.
- Remove `d-flex` and `justify-content-end` classes from configuration
- Open the dropdown
- Check that the dropdown has the correct position by checking that it has the inline style left and not right.

These steps would cover the 2 tickets, and we would simplify everything into a single test. What do you think?

File: modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js (hunk 1 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-collection-contributor/fragment-collection-contributor-basic-component/src/main/resources/com/liferay/fragment/collection/contributor/basic/component/dependencies/dropdown/index.js:173]: Thanks for the inputs on this, you are right.

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts (hunk 2 above)
Comment (georgel-pop-lr) [reply modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:43]: Yes, it sounds good.

Comment (georgel-pop-lr) [reply modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts]: @veroglez thanks for the review, I did the changes, please check. 

Regards.

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+	// Open dropdown
+
+	await dropdownButton.press(ENTER_KEY);
+
+	const dropdownMenu = await dropdownFragment.locator(
+		'.dropdown-fragment-menu'
+	);
+
+	await dropdownMenu.getByText('Place fragments or widgets here.').waitFor();
```
Comment (veroglez) [inline modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:51]: instead, maybe we can use this?
``` 
await dropdownMenu.waitFor({state: 'visible'});
```

since a few lines later we use this same check but with `hidden`

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			pageDefinition: getPageDefinition([fragmentDefinition]),
+			siteId: site.id,
+			title: getRandomString(),
+		});
+
+		await pageEditorPage.goto(layout, site.friendlyUrlPath);
+
+		await pageEditorPage.selectFragment(dropdownId);
+
+		const dropdownFragment = await pageEditorPage.getFragment(dropdownId);
```
Comment (veroglez) [inline modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:93]: I think the followings lines are not necessary

```
const dropdownFragment = await pageEditorPage.getFragment(dropdownId);
const dropdownButton = await dropdownFragment.locator('.dropdown-fragment-toggle');

await expect(dropdownButton.getByText('Dropdown', {exact: true})).toBeVisible();

```


we can check directly with the method `openDropdownAndCheckStyle` in desktop viewport

```
await openDropdownAndCheckStyle(
	pageEditorPage.getFragment(dropdownId),
	'right',
	'280px'
);
```

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		const dropdownButton = await dropdownFragment.locator(
+			'.dropdown-fragment-toggle'
+		);
+
+		await expect(
+			dropdownButton.getByText('Dropdown', {exact: true})
+		).toBeVisible();
+	});
+
+	await test.step('Check position in desktop and mobile viewport', async () => {
```
Comment (veroglez) [inline modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:104]: We do not usually use the `test.step` pattern since sometimes the code can become longer and more difficult to read, if we have to clarify something we type a small comment. but I don't have a strong opinion on this, so I leave it to your choice :slightly_smiling_face:

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+		for (const [viewport, value] of Object.entries(VIEWPORTS_VALUES)) {
```
Comment (veroglez) [inline modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:105]: I think this `for` makes it a little difficult to read the code. I think that for cases in which we have few elements, (like this case, where there are only 2) I prefer to make the checks explicit for both elements and avoid the loop pattern. I think the code is more readable and this way we also simplify the logic to see if we are in destktop layout or not. What do you think? :smile:

File: modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+);
+
+const VIEWPORTS_VALUES = {
+	'Desktop': '280px',
+	'Portrait Phone': '0px',
+};
+
+const ENTER_KEY = 'Enter';
+
+async function checkStyleDropdownMenu(
```
Comment (veroglez) [inline modules/test/playwright/tests/layout-content-page-editor-web/dropdownFragment.spec.ts:34]: since this method opens the dropdown and checks the css, we could call it `openDropdownAndCheckStyle` to make it more readable. What do you think?

File: modules/test/playwright/pages/layout-content-page-editor-web/PageEditorPage.ts [hunk 8]
```diff
@@ -201,7 +208,16 @@ export class PageEditorPage {
 					return;
 				}
 
-				await field.fill(value as string);
+				if (Array.isArray(value)) {
```
Comment (veroglez) [inline modules/test/playwright/pages/layout-content-page-editor-web/PageEditorPage.ts:211]: I think you don't use this method in the test, right? Why did you change it?

**Comment by georgel-pop-lr** [comment]
ci:test:sf

**Comment by georgel-pop-lr** [comment]
ci:test:sf

**Comment by veroglez** [comment]
ci:test:echo-playwright

**Comment by ealonso** [comment]
ci:forward

**Comment by ealonso** [comment]
ci:forward

**Comment by ealonso** [comment]
ci:forward:force
